internal/webhook/external/artifactory.go

```diff
@@ -0,0 +1,196 @@
+package external
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/helm"
+	xhttp "github.com/akuity/kargo/internal/http"
+	"github.com/akuity/kargo/internal/image"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	artifactorySecretDataKey = "secret-token"
+	artifactory              = "artifactory"
+)
+
+func init() {
+	registry.register(
+		artifactory,
+		webhookReceiverRegistration{
+			predicate: func(cfg kargoapi.WebhookReceiverConfig) bool {
+				return cfg.Artifactory != nil
+			},
+			factory: newArtifactoryWebhookReceiver,
+		},
+	)
+}
+
+// artifactoryWebhookReceiver is an implementation of WebhookReceiver that handles
+// inbound webhooks from Jfrog Artifactory.
+type artifactoryWebhookReceiver struct {
+	*baseWebhookReceiver
+}
+
+// newArtifactoryWebhookReceiver returns a new instance of artifactoryWebhookReceiver.
+func newArtifactoryWebhookReceiver(
+	c client.Client,
+	project string,
+	cfg kargoapi.WebhookReceiverConfig,
+) WebhookReceiver {
+	return &artifactoryWebhookReceiver{
+		baseWebhookReceiver: &baseWebhookReceiver{
+			client:     c,
+			project:    project,
+			secretName: cfg.Artifactory.SecretRef.Name,
+		},
+	}
+}
+
+// GetDetails implements WebhookReceiver.
+func (a *artifactoryWebhookReceiver) getReceiverType() string {
+	return artifactory
+}
+
+// getSecretValues implements WebhookReceiver.
+func (a *artifactoryWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	secretValue, ok := secretData[artifactorySecretDataKey]
+	if !ok {
+		return nil,
+			errors.New("Secret data is not valid for a Artifactory WebhookReceiver")
+	}
+	return []string{string(secretValue)}, nil
+}
+
+// getHandler implements WebhookReceiver.
+func (a *artifactoryWebhookReceiver) getHandler(requestBody []byte) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+
+		token, ok := a.secretData[artifactorySecretDataKey]
+		if !ok {
+			xhttp.WriteErrorJSON(w, nil)
+			return
+		}
+
+		sig := r.Header.Get("x-jfrog-event-auth")
+		if sig == "" {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					errors.New("missing signature"),
+					http.StatusUnauthorized,
+				),
+			)
+			return
+		}
+
+		mac := hmac.New(sha256.New, token)
+		mac.Write(requestBody)
+		computedSig := hex.EncodeToString(mac.Sum(nil))
+
+		if !hmac.Equal([]byte(sig), []byte(computedSig)) {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					errors.New("unauthorized"),
+					http.StatusUnauthorized,
+				),
+			)
+			return
+		}
```

Comment:
Would the signature validation from GitHub, which was generic enough that we've reused it in some other receivers, work here as well?

Note that I have not actually tried this:

```suggestion
		// Note: github.com/google/go-github/v71/github has a great implementation
		// of HMAC signature validation that isn't GitHub-specific, so we've opted
		// to use it here for Artifactory as well.
		if err := gh.ValidateSignature(sig, requestBody, signingKey); err != nil {
			xhttp.WriteErrorJSON(
				w,
				xhttp.Error(errors.New("unauthorized"), http.StatusUnauthorized),
			)
			return
		}
```